python/ray/experimental/data/dataset.py

```diff
@@ -314,7 +388,12 @@ def schema(self) -> Union[type, "pyarrow.lib.Schema"]:
         Returns:
             The Python type or Arrow schema of the records.
         """
-        raise NotImplementedError  # P0
+        metadata = self._blocks.get_metadata()
```

Comment:
maybe we need a todo task to validate that all the blocks has same schema to avoid surprises